src/components/evaluate.ts

```diff
@@ -0,0 +1,33 @@
+const expr = require('expression-eval');
```

Comment:
Aha! Just for funsies, here's a quick fix:

```js
const compile = context => expr => Function('s', `with(s) { return ${expr}; }`)
  .bind(context)(new Proxy(context, ...;
```

and then here's how to break out again:

```js
compile({a: 5, this: 5})("(function(){}).constructor('alert(window)')()")
```

Reply:
Yeah, I have a more robust version of it somewhere, but it has around 100 lines of code , so it's not smaller than expresion-eval. :laughing: and still not safe, as you could do sth like `[].prototype.constructor` (I didn't make a copy of everything).

Hopefully, realms land in spec soon.

Re expression-eval... Could we not use it? I see they don't filter out CallExpressions and MemberExpresions. We seem to need it only for string concatenation or simple addition, right?

If so, why don't we simply pull BinaryExpression?

```
function getLiterals(node: any): Array<number | string> {
  switch (node && node.type) {
    case 'BinaryExpression':
      return [...getLiterals(node.left), ...getLiterals(node.right)];
    case 'Literal': // no StringLiteral and NumericLiteral :(
      if (/^['"]/.test(node.raw)) {
        // string
        return [node.value];
      }

      return [Number(node.value)];
    default:
      return [];
  }
}

getLiterals(jsep('2 + 4 + \'2\''));
```
We could also add support for ConditionalExpression or UnaryExpression if needed.

Reply:
@XVincentX thoughts? ^
You know the use case better than me, I believe.

Reply:
This is a complicated call for me.

On a side, smaller the surface is, more unlikely there's going to be a security flaw. 

On the other side, I think in general we're (we as us in general) not good at security, so I'd rather rely on other libraries rather than figuring out it by ourselves.

I'd be happy if we could rely on some sandbox that's battle tested in the wild; I was suggesting last week privately to @wmhilton to evaluate https://github.com/nx-js/compiler-util, although it seems to be overpowered for us.

Reply:
Yeah, it might be overkill, but if we need to allow JS execution, this is the only way to go unless we want to make use of sandboxed iframes + some separate thing (node's vm module?) for electron.
Let's wait for Will then and evaluate what kind of expressions need to be supported.

Reply:
I think it's fine as-is. It seems to be rather difficult in practice to use functions inside `expression-eval`. For instance, none of these worked:

```json
"evalOptions": "[Math.max(0, 2)]",
"evalOptions": "['hello'.toUpperCase()]",
"evalOptions": "[(function(x) { return x.toUpperCase() })('hello')]",
"evalOptions": "['int32','int64'].map(function(x) { return x.toUpperCase() })",
"evalOptions": "['int32','int64'].map(x => x.toUpperCase())",
"evalOptions": "[(function(x){ return x + x })('int32')]",
```

In fact, as far as I can tell, the only way you can "call" functions is if you passed them in as part of the context. E.g.

```js
evaluate("console.log('hi')", { console });
```

And since our `context` is just a key/value map of the current form values, there aren't any functions exposed. But it's nice that it allows that, because if the future we could add helper functions like `sort` or something.

Remember, at the end of the day we're writing both the schemas and the Formtron components. Therefore, both the functions and the context are under our control. And what's the worst that happens if a "malicious" user does eval something? They could do much worse simply by opening the Developer Console. I think there's basically zero risk in terms of security threat model.